src/models/transfers.js

```diff
@@ -253,16 +256,18 @@ function * fulfillTransfer (transferId, fulfillment) {
     const accountBalances = yield makeAccountBalances(transaction, transfer)
 
     if (isExecution) {
-      if (transfer.state !== 'prepared') {
+      if (!_.includes(validExecutionStates, transfer.state)) {
         throw new InvalidModificationError('Transfers in state ' +
           transfer.state + ' may not be executed')
       }
-      yield accountBalances.applyCredits()
+      if (transfer.state === 'prepared') {
+        yield accountBalances.applyCredits()
+      }
       updateState(transfer, 'executed')
```

Comment:
To be safe, we might need to look up the existing fulfillment because for example there may be multiple fulfillments that satisfy the condition and we should probably be returning the first fulfillment since the second one has no actual effect. So after `validateConditionFulfillment`, if the transfer state is `execution`, just look up the existing fulfillment and return like the code in your comment.